### PR TITLE
Add extending extensions tuturial based on imkingdavids topics

### PR DESCRIPTION
**Pull request by paul999**
Based on the original documentation from @imkingdavid , in this topic https://www.phpbb.com/community/viewtopic.php?f=461&t=2210001

### CODE REVIEW

**Comment by paul999** [comment]
@VSEphpbb @marc1706

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 1]
```diff
@@ -0,0 +1,169 @@
+========================================================
+Tutorial: Letting other extensions extend your extension
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: I prefer "Allowing" instead of "Letting"

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+This tutorial is based on the work by `imkingdavid <https://www.phpbb.com/community/viewtopic.php?f=461&t=2210001>`_
+
+Introduction
+============
+
+This tutorial explains:
+
+* Adding a PHP event to your extension
+* Adding template event to your extension
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: Adding `a` template event to your extension (here and in the title below)

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+* Using service collection
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: collections (here and below)

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+* Using the phpBB finder tool
+
+Adding a PHP event to your extension
+====================================
+The first of the three is probably the easiest, but is limited in abilities once
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: three? I count four. I would replace `three` with `these`

Comment (paul999) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: Nha, your counting is off :P (I blame Micheal  :P)

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  list of events provided by your extension and make public note of any event
+  changes prior to releasing a new version.
+
+The name of your event should be in the form of acme.demo.identifer, where
+vendor.name matches the value in your composer.json.
+
+Adding template event to your extension
+=======================================
+Besides PHP events there are also template events. Template events can be used
+to extent the template of a extension. To create a template event you simply add
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: `extent` -> `extend`

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+service.
+
+.. code-block:: yaml
+
+  acme.demo.foobar_manager:
+      arguments:
+          - '@acme.demo.foobar_collection'
+
+That argument will return an instance of ``phpbb\di\service_collection``, which
+extends ``ArrayObject`` so it can be used an array containing the service name of
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: so it can be used `as` an array...

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+The ``\phpbb\extension\finder`` is available from the service container as
+``ext.finder`` and can be used as follows. The following example is part of what
+is used to find all routing files, both for extensions and core routes.
+
+.. code-block:: php
+
+  $finder
+      ->directory('config')
+      ->suffix('routing.yml')
+      ->find()
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: missing semi-colon ;P

**Comment by paul999** [comment]
Updated :)

**Comment by imkingdavid** [comment]
Oooh now I feel all special. :P Feel free to update this if anything is out of date since I wrote it in 2013. xD

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+As you can see, you are able to chain method calls together (ending with ``find()``).
+Check out the class definition for more information about the different methods
+that are available (such as specifying a different directory for extensions than
+core files, and getting class names based on files it finds). The return of the
+``find()`` method is an array of file paths that match the given criteria.
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: I would also add maybe this:

```
For example, assume you have an extension and you need to get a collection of all PNG image files from one of its directories. You could use the extension manager to load the finder, and traverse the extension's image directory as follows:

.. code-block:: php

  $finder = $extension_manager->get_finder();

  $images = $finder
      ->extension_suffix('.png')
      ->extension_directory('/images')
      ->find_from_extension('demo', $phpbb_root_path . 'ext/acme/demo/');

The $images array would look something like (as you can see the image paths are contained in the array keys):

.. code-block:: php

  array(
      'ext/acme/demo/images/image1.png' => 'demo',
      'ext/acme/demo/images/image2.png' => 'demo',
  );
```

**Comment by Nicofuma** [comment]
I'd like to add a section about service decoration (only available in 3.2.x) and service replacement somewhere. 

About service replacement the goal is to say that we can't assume the order and it is generally a bad idea (because if 2 extensions are doing the same the will conflict).
About service decoration, it remove the necessity of events at the beginning or the end of functions.

**Comment by paul999** [comment]
Will add something regarding it tomorrow evening.

**Comment by VSEphpbb** [comment]
Maybe it's also worth adding something about the ordered service collections too, in 3.2.

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+Using service collections
+=========================
+In 3.1, a new concept is that of "services". You can read up on exactly what a
+service is `here <http://symfony.com/doc/current/book/service_container.html>`_;
+the rest of this guide will assume you have basic knowledge of services and how
+to use them (if you have further questions, feel free to ask). A service
+collection is basically what it sounds like: a collection of services. Basically,
+when you define your service, you give it a special "tag", which associates it
+with a the collection of services. Later on, this can be used to easily get a
```
Comment (rxu) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: ```with a the```

**Comment by paul999** [comment]
Updated to include some more stuff, @VSEphpbb @Nicofuma

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 10]
```diff
@@ -140,6 +140,25 @@ as the corresponding value.
 This system is used in the core for several features, including notifications
 and authentication providers.
 
+Ordered service collection
+--------------------------
+.. note::
+  Ordered service collections are introduced in phpBB 3.2.
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: are -> were

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+
+This system is used in the core for several features, including notifications
+and authentication providers.
+
+Ordered service collection
+--------------------------
+.. note::
+  Ordered service collections are introduced in phpBB 3.2.
+
+Ordered service collections are based on a normal service collection, but the
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: I would start with something like this sentence too:
Ordered service collections allow you to define the order in which services are loaded, which is especially useful in cases where service priority and/or dependency requires they be loaded in a specified order.

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+except instead of using ``service_collection`` you should use ``ordered_service_collection``:
+
+.. code-block:: yaml
+
+  acme.demo.foobar_collection:
+      class: phpbb\di\ordered_service_collection
+      arguments:
+          - '@service_container'
+      tags:
+          - { name: service_collection, tag: acme.demo.foobar_service }
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]:
```
    acme.demo.foobar_foo:
        class: acme\demo\foobar_foo
        tags:
            - { name: acme.demo.foobar_service, order: 1 }

    acme.demo.foobar_bar:
        class: acme\demo\foobar_bar
        tags:
            - { name: acme.demo.foobar_service, order: 2 }
```

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 13]
```diff
@@ -0,0 +1,305 @@
+=========================================================
+Tutorial: Allowing other extensions extend your extension
+=========================================================
+
+Some parts of this tutorial is based on the work by `imkingdavid <https://www.phpbb.com/community/viewtopic.php?f=461&t=2210001>`_
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: is -> are

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+=========================
+.. warning::
+  Only use service replacements if your extensions needs to be compatible with
+  phpBB 3.1. For 3.2 and later, see `Using service decoration`_.
+
+.. warning::
+  You can't assume the order of a service is defined in phpBB if other extensions
+  are installed. Be really carefully when using service replacements.
+
+With a service replacement you can replace a existing service in phpBB (Or in an
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: a existing -> an existing

Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: Or -> or

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+extension) with your own service. Your replacement should type match the original
+service (so if the original service implements a interface, you should minimal
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: `an` interface
minimal -> `at least` or `at a minimum`

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 16]
```diff
@@ -0,0 +1,305 @@
...
+implement that specific interface, if it is a concrete class, you will need to
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: replace comma with period between interface and if.

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+extent that class). The best way to accomplish this is by extending the original
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: extent -> extend

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+class, and only replace the features you want to change.
+
+To replace a core phpBB service, you simply name your extensions service name
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: extensions -> extension's

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 19]
```diff
@@ -0,0 +1,305 @@
...
+with the same name as the service in phpBB's core.
+
+If you want, for example, replace the ``config`` service in phpBB with your own
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: Reword it to -> For example, to replace the config...

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 20]
```diff
@@ -0,0 +1,305 @@
...
+implementation, your service configuration will look like this:
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: will -> would

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 21]
```diff
@@ -0,0 +1,305 @@
...
+
+.. code-block:: yaml
+
+  config:
+      class: acme\demo\config\db
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: why `\db` at the end?

Comment (paul999) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: Good one. I guess I based it on \phpbb\config\db as that is the base clase for phpBBs service definition.

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 22]
```diff
@@ -0,0 +1,305 @@
...
+      arguments:
+          - '@dbal.conn'
+          - '@cache.driver'
+          - '%tables.config%'
+          - '@acme.demo.db_reader'
+
+Your php class should extends, at least, ``\phpbb\config\config``, which is the base
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: Your PHP class could extend ``\phpbb\config\config`` which is the base class phpBB extends from. However, to avoid any unforseen problems, it's better to more explicitly extend from the ``\phpbb\config\db`` class, which is the specific class being replaced by the service.

File: development/extensions/tutorial_php_events_and_extending_extensions.rst (hunk 21 above)
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: nvmd, I see now

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 23]
```diff
@@ -0,0 +1,305 @@
...
+class phpBB extends from. However, to be sure everything still works, you should
+extend from ``\phpbb\config\db``, which is the original class the service had.
+
+.. warning::
+  If you are using EPV in travis, or during submission toe the extensions database
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: toe ?

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 24]
```diff
@@ -0,0 +1,305 @@
...
+  at phpBB.com, you will receive a warning that your service configuration
+  doesn't follow the extensions database policies. As you are overwriting a core
+  service, you can simply ignore this message. However, in all cases you should
+  inform the the phpBB extensions team why you got the warning.
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: got -> received

File: development/extensions/tutorial_php_events_and_extending_extensions.rst (hunk 23 above)
Comment (paul999) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: there should be at least 1 dutch word in every sentence :P

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 25]
```diff
@@ -0,0 +1,305 @@
...
+
+Using service decoration
+========================
+.. note::
+  See for full documentation regarding service decoration the
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: Read about Service Decoration at Symphony for complete documentation.

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 26]
```diff
@@ -0,0 +1,305 @@
...
+  `Symfony <http://symfony.com/doc/current/components/dependency_injection/advanced.html#decorating-services>`_
+  documentation.
+
+Instead of replacing a service, you should decorate the service. The will
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: From phpBB 3.2, you can use service decoration as the preferred method to replace existing services, in the core or from other extensions. Decoration will update an existing service with a new name, leaving it intact so that it can be referenced in the new service.

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 27]
```diff
@@ -0,0 +1,305 @@
...
+make the original service being renamed to a new name, so the old service is
+kept around and can be referenced in the new service.
+
+Decorating the original service is done by setting the decororates option in your
+service.yml:
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: For example, to replace `config` with `acme.demo.decorated`, simply add the `decorates` option to its service definition:

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 28]
```diff
@@ -0,0 +1,305 @@
...
+
+.. code-block:: yaml
+
+  acme.demo.decorated:
+    class: '\acme\demo\decorated\config'
+    decorates: 'config'
+
+If you want to inject the original service, which is renamed to have .inner appended
+to its name, you can use:
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: In Symphony, the old `config` service will have been renamed `config.inner`, so you can inject it into your new service by adding it to your services arguments:

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 29]
```diff
@@ -0,0 +1,305 @@
...
+    class: '\acme\demo\decorated\config'
+    decorates: 'config'
+    arguments:
+      - '@config.inner'
+
+Again, keep in mind that your new class type matches the original class.
+
+Replacing the phpBB Datetime class
+==================================
+If you want to replace the phpBB Datetime class, for example if you want to use
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: remove second `if you want` so it says `for example to use...`

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 30]
```diff
@@ -0,0 +1,325 @@
...
+    /**
+    * Short description of the event
+    *
+    * @event acme.demo.identifier
+    * @var  string  var1  A variable to make available to the event
+    * @var  string  var2  A second variable to make available to the event
+    * @since 3.1-A1
+    */
+    $vars = array('var1', 'var2');
+    extract($phpbb_dispatcher->trigger_event('acme.demo.identifier', compact($vars)));
```
Comment (Nicofuma) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: I wonder if we should talk about the dispatch() method and the ability of using a custom class even here... maybe in a `note` block or something like that with a link to the related symfony documentation?

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 31]
```diff
@@ -0,0 +1,325 @@
...
+
+  <!-- EVENT acme_demo_identifier -->
+
+The event can be used in the same way as the template events in the phpBB Core.
+See :doc:`tutorial_basics` on how to use these events.
+
+.. warning::
+  Like with PHP events you should not change the identifier of the event after
+  a release of your extension. Other extensions might use your event and be broken
+  afterwards.
```
Comment (Nicofuma) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: I would add a note block with the twig syntax: `{% EVENT acme_demo_identifier %}`

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 32]
```diff
@@ -0,0 +1,325 @@
...
+to load the finder, and traverse the extension's image directory as follows:
+
+.. code-block:: php
+
+  $finder = $extension_manager->get_finder();
+
+  $images = $finder
+      ->extension_suffix('.png')
+      ->extension_directory('/images')
+      ->find_from_extension('demo', $phpbb_root_path . 'ext/acme/demo/');
```
Comment (Nicofuma) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: `find_from_extension` will search for the files in a specific extension. In 99% of the use cases i think we will prefer `find()` which will search in all extensions.

In 99% of the use case, `suffix()` and `directory()` ill be enough too (they are used when we need to specify different directory or suffix for the core and the extensions. By the way,  if we want to search in the core we need to call core_path() too)

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 33]
```diff
@@ -0,0 +1,325 @@
...
+
+The ``$images`` array would look something like (as you can see the image paths are
+contained in the array keys):
```
Comment (Nicofuma) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: I would also specify that the value is the name of the extension containing the file

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 34]
```diff
@@ -0,0 +1,325 @@
...
+extends from. However, to avoid any unforseen problems, it's better to more
+explicitly extend from the ``\phpbb\config\db`` class, which is the specific class
+being replaced by the service.
+
+.. warning::
+  If you are using EPV in travis, or during submission to the extensions database
+  at phpBB.com, you will receive a warning that your service configuration
+  doesn't follow the extensions database policies. As you are overwriting a core
+  service, you can simply ignore this message. However, in all cases you should
+  inform the the phpBB extensions team why you received the warning.
```
Comment (Nicofuma) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: double the

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 35]
```diff
@@ -0,0 +1,325 @@
...
+      arguments:
+          - '@dbal.conn'
+          - '@cache.driver'
+          - '%tables.config%'
+          - '@acme.demo.db_reader'
+
+Your PHP class could extend ``\phpbb\config\config`` which is the base class phpBB
+extends from. However, to avoid any unforseen problems, it's better to more
+explicitly extend from the ``\phpbb\config\db`` class, which is the specific class
+being replaced by the service.
```
Comment (Nicofuma) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: To be fair, in this specific case the contract is `config\config` not `config\db` and I think it could be a good idea to provide an example doing that (while completely changing the services arguments)

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 36]
```diff
@@ -0,0 +1,325 @@
...
+For example, to replace ``config`` with ``acme.demo.decorated.config``, simply
+add the ``decorates`` option to its service definition:
+
+.. code-block:: yaml
+
+  acme.demo.decorated:
+    class: '\acme\demo\decorated\config'
+    decorates: 'config'
+
+In Symfony, the old config service will have been renamed config.inner, so you
```
Comment (Nicofuma) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: actually I think it is  `acme.demo.decorated.inner`

File: development/extensions/tutorial_php_events_and_extending_extensions.rst (hunk 33 above)
Comment (paul999) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: @Nicofuma Do you mean the key index for the array is based on the vendor name? It is kinda implicit from the example I think?

File: development/extensions/tutorial_php_events_and_extending_extensions.rst (hunk 32 above)
Comment (paul999) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: I added a small example in a note, but not fully sure it is correct, can you check @Nicofuma  ?

File: development/extensions/tutorial_php_events_and_extending_extensions.rst (hunk 31 above)
Comment (paul999) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: Added.

File: development/extensions/tutorial_php_events_and_extending_extensions.rst (hunk 30 above)
Comment (paul999) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: Hmm, I am not fully sure. I think this is already going pretty deep

**Comment by paul999** [comment]
And updated again with mostly fixes from @Nicofuma

File: development/extensions/tutorial_php_events_and_extending_extensions.rst (hunk 33 above)
Comment (Nicofuma) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: that's rigth

File: development/extensions/tutorial_php_events_and_extending_extensions.rst (hunk 30 above)
Comment (Nicofuma) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: ok that's fine, by the way I don't remember. Is there a note or something else with a link to Symfony's event dispatcher documentation?

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 37]
```diff
@@ -231,6 +238,19 @@ contained in the array keys):
...
+  ``directory`` and ``find`` instead:
+
+  .. code-block:: php
+
+    $finder = $extension_manager->get_finder();
+
+    $images = $finder
+        ->find('.png')
+        ->directory('/images')
+        ->find($phpbb_root_path . 'ext/');
```
Comment (Nicofuma) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: IIRC it should be `/` in our case (i.e: it wil look from the root directory of the core and of each extension)

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 38]
```diff
@@ -231,6 +238,19 @@ contained in the array keys):
       'ext/acme/demo/images/image2.png' => 'demo',
   );
 
+.. note::
+  The method ``find_from_extension`` used above will only search in that specific
+  extension. If you want to search for images in all extensions, you use ``suffix``,
+  ``directory`` and ``find`` instead:
```
Comment (Nicofuma) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: That's a bit more complicated:

 - find() will search in both core and extensions depending on what is configured (if a path is configured for the core it will search in core and extensions, otherwise only extensions)
 - extension_*() methods will configure the options for the extensions only
 - core_*() methods will configure the options for the core only
 - suffix/prefix/...() methods will configure the options for both (options, not path)
 - core_path() will configure the path for the core

File: development/extensions/tutorial_php_events_and_extending_extensions.rst (hunk 37 above)
Comment (Nicofuma) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: Actually find() doesn't take any argument of this kind. Only $cache and $is_dir ($is_dir is to returns files and directories)

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 39]
```diff
@@ -0,0 +1,344 @@
...
+      class: acme\demo\config\db
+      arguments:
+          - '@auth'
+          - '@passwords.manager'
+          - '@acme.demo.db_reader'
+
+Your PHP class should extend ``\phpbb\config\db`` which is original class in
+phpBB. To avoid any unforseen problems, you should not extend directly
+from ``\phpbb\config\config``. If both implemented the same interface, you could
+have extended from ``\phpbb\config\config``.
```
Comment (Nicofuma) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: I'm not sure to understand what you want to say here.

In order to replace the config service, the class must extends `\phpbb\config\config` or any of its subclasses. That's the contract here (but not easy to guess as there isn't any interface...). So considering that there isn't any interface I think your statement is correct.

File: development/extensions/tutorial_php_events_and_extending_extensions.rst (hunk 30 above)
Comment (paul999) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: No, but I will add one.

File: development/extensions/tutorial_php_events_and_extending_extensions.rst (hunk 37 above)
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: That code is taken directly from current working code.

Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]:
```php
$images = $finder
	->suffix('.png')
	->core_path('ext/')
	->find();
```

Comment (Nicofuma) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: no core_path call, that's used only for looking in core files

**Comment by paul999** [comment]
@Nicofuma @VSEphpbb  Updated again. Can we get it merged now?

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 40]
```diff
@@ -284,10 +284,12 @@ implementation, your service configuration would look like this:
           - '@passwords.manager'
           - '@acme.demo.db_reader'
 
-Your PHP class should extend ``\phpbb\config\db`` which is original class in
-phpBB. To avoid any unforseen problems, you should not extend directly
-from ``\phpbb\config\config``. If both implemented the same interface, you could
-have extended from ``\phpbb\config\config``.
+The original config class in ``\phpbb\config\db`` didn't implement an interface.
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: The original config class in ``\phpbb\config\db`` does not implement an interface.

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 41]
```diff
@@ -284,10 +284,12 @@ implementation, your service configuration would look like this:
...
+This means you need to extend the original ``\phpbb\config\db``, otherwise type
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: This means you need to extend the original ``\phpbb\config\db``, otherwise the type

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 42]
```diff
@@ -284,10 +284,12 @@ implementation, your service configuration would look like this:
...
+won't match the type hinting in the constructors which use the config service.
+If the original service implemented a interface directly, and all type hints
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: If the original service implements an interface directly, and all type hints

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 43]
```diff
@@ -284,10 +284,12 @@ implementation, your service configuration would look like this:
...
+are done based on the interface, you are not required to extend the original class
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: are referencing the interface, you are not required to extend the original class

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 44]
```diff
@@ -284,10 +284,12 @@ implementation, your service configuration would look like this:
...
+but you can simply implement the interface.
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: and should instead implement the interface.

File: development/extensions/tutorial_php_events_and_extending_extensions.rst [hunk 45]
```diff
@@ -240,17 +240,23 @@ contained in the array keys):
...
-        ->directory('/images')
-        ->find($phpbb_root_path . 'ext/');
+        ->core_path('ext/')
+        ->find();
+
+.. note::
+  Depending on how the Finder class is configured, the find method searches both
+  the phpBB core files as the extensions directories. You can use the extension_*
+  methods in the Finder class to configure extension specific search, and you can
+  use the core_* methods to specific core specific searches.
```
Comment (VSEphpbb) [inline development/extensions/tutorial_php_events_and_extending_extensions.rst]: use the core_* methods to configure core specific searches.

**Comment by Nicofuma** [comment]
Nice :+1: (after VSE is happy with the wording :))

**Comment by paul999** [comment]
Fixed all of them @VSEphpbb  :dancer:

**Comment by VSEphpbb** [comment]
👍

**Comment by paul999** [comment]
👯 👯 👯 👯 👯 👯 👯
